village/src/App.js

```diff
@@ -1,27 +1,135 @@
   // You'll need to make sure you have the right properties on state and pass them down to props.
+
+  componentDidMount() {
+    axios
+      .get("http://localhost:3333/smurfs")
+      .then(response => {
+        this.setState({
+          smurfs: response.data
+        });
+      })
+      .catch(err => console.log(err));
+  }
+  //this makes a call to our server and uses the dataset from the response to set our state.
+
+  postSmurf = newSmurf => {
+    axios
+      .post("http://localhost:3333/smurfs", newSmurf)
+      .then(response => {
+        this.setState({
+          smurfs: response.data,
+          selectedSmurf: ""
+        });
+      })
+      .catch(err => console.log(err));
+  };
+  //this sends our new smurf data as an argument to post to the server. It then sets state on App to update our state with the new smurf added.
+
+  deleteSmurf = id => {
+    axios
+      .delete(`http://localhost:3333/smurfs/${id}`)
+      .then(response => {
+        this.setState({
+          smurfs: response.data,
+          selectedSmurf: ""
+        });
+      })
+      .catch(err => console.log(err));
+  };
+  //this takes an id from Smurfs and sends a delete request to the server for the smurf matching that id. Then updates our state with the new data.
+
+  putSmurf = updatedSmurf => {
+    axios
+      .put(
+        `http://localhost:3333/smurfs/${this.state.selectedSmurf.id}`,
+        updatedSmurf
+      )
+      .then(response => {
+        this.setState({
+          smurfs: response.data,
+          selectedSmurf: ""
+        });
+      })
+      .catch(err => console.log(err));
+  };
+
+  selectSmurf = smurf => {
+    this.setState({
+      selectedSmurf: smurf
+    });
+  };
+
   render() {
+    if (this.state.smurfs.length === 0) {
+      return <div>Loading...</div>;
+    }
     return (
       <div className="App">
-        <SmurfForm />
-        <Smurfs smurfs={this.state.smurfs} />
+        <header>
+          <NavLink exact to="/">
+            Smurf List
+          </NavLink>
+          <NavLink to="/smurf-form">Modify Smurfs</NavLink>
+        </header>
+        <Route
+          exact
+          path="/smurf-form"
+          render={props => (
+            <SmurfForm
+              {...props}
+              selectedSmurf={this.state.selectedSmurf}
+              postSmurf={this.postSmurf}
+              putSmurf={this.putSmurf}
+            />
+          )}
+        />
+        <Route
+          path="/"
+          render={props => (
+            <Smurfs
+              {...props}
+              smurfs={this.state.smurfs}
+              deleteSmurf={this.deleteSmurf}
+              selectSmurf={this.selectSmurf}
+            />
+          )}
+        />
+        <Route
+          exact
+          path="/smurf/:id"
+          render={props => (
+            <DirectSmurf
+              {...props}
+              linkedSmurf={this.state.smurfs.find(smurf => {
+                return smurf.id === this.props.match.params.id;
```

Comment:
this.props.match.params here is going to refer to App's match.params, not DirectSmurf's match.params, which is why this line isn't doing what you want it to